x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/scheduler/SchedulerEngine.java

```diff
@@ -136,6 +137,10 @@ public void stop() {
         }
     }
 
+    public Set<String> scheduledJobIds() {
+        return schedules.keySet();
```

Comment:
Yep, I'll change this.

Reply:
Switching this to `Collections.unmodifiableSet(schedules.keySet());` doesn't quite go far enough - the returned collection can't be modified, which solves that problem, but it's just a view into the underlying collection, not a copy, so iterating over it while removing items still hits the same undefined case.  Whether we want to make a copy in this method or at the site of the consumer I'm not sure - I'd err on the side of caution and do it here, but I could see the other way too.

Reply:
> iterating over it while removing items still hits the same undefined case

Not sure I follow here, can you clarify what you mean? From the `unmodifiableSet` javadocs:

> Query operations on the returned set "read through" to the specified set, and attempts to modify the returned set, whether direct or via its iterator, result in an `UnsupportedOperationException`.

Reply:
Right, I'm looking at [`cancelSnapshotJobs()`](https://github.com/elastic/elasticsearch/pull/39795/files#diff-65a7e37d0dc6108c8613c07d4315ec9eR102), which calls `scheduledJobIds()`, then iterates over the returned set (which is a view into `schedules`), calling `remove()` on each one, which removes it from `schedules`.

So we're iterating over an unmodifiable view into `schedules` while removing things from `schedules`, which is exactly this case from the `keySet()` documentation: `If the map is modified while an iteration over the set is in progress (except through the iterator's own remove operation), the results of the iteration are undefined.`.

Fundamentally the problem here is that `Collections.unmodifiableX` is (unintuitively) just a view, not a copy - the returned instance can't be used to modify it, but if something else is holding on to the original object, it can be mutated and the "unmodifiable" instance will change.

E.g.:
```
        Map<String, String> test = new ConcurrentHashMap<>();
        test.put("key1", "val1");
        test.put("key2", "val2");
        Set<String> immutableTest = Collections.unmodifiableSet(test.keySet());
        test.remove("key1");

        System.out.print("test: ");
        System.out.println(test);

        System.out.print("immutableTest: ");
        System.out.println(immutableTest);
```
prints:
```
test: {key2=val2}
immutableTest: [key2]
```

Reply:
Ahh okay, I thought you were referring to something like this:

```java
        Iterator<String> i = scheduler.scheduledJobIds().iterator();
        while (i.hasNext()) {
            i.remove(); // <-- this "remove"
            i.next();
        }
```

Which is prevented, rather than the `remove` on the `SchedulerEngine`'s internal map.

I'll just copy the set.